### DISCUSSION TITLE
Integrate with Google Classroom

### DISCUSSION THREAD

**Discussion by tachyons** [CircuitVerse/CircuitVerse#2714] [Ideas]
We have LMS integrations with LTI tool. But Google Classroom is not compatible with LTI and need to integrated separately.

**Comment by tachyons** [comment]
Previous attempts for ref:

https://www.youtube.com/watch?v=7svs_j8MPA0 from @koolgax99

**Comment by pulkit-30** [comment]
Hey @tachyons,
 I would like to work on this issue. May you Please assign me this issue.

**Comment by tachyons** [comment]
@Prerna-0202 Asked for this first, you can take in case she isn't taking this

**Comment by Prerna-0202** [comment]
@pulkit-30, yes you can start working on this issue...

**Comment by pulkit-30** [comment]
@Prerna-0202 @tachyons, I am starting working on this.

**Comment by koolgax99** [comment]
@pulkit-30  https://github.com/koolgax99/CircuitVerse-GSoC-Task you could probably check out this repo for your ref. The only problem is how u share the circuit link with the gclassroom's api

**Comment by tachyons** [comment]
@ayan-biswas0412 Also did some work on this

**Comment by koolgax99** [comment]
> @ayan-biswas0412 Also did some work on this

Yes yes we did it as a gsoc task.😁

**Comment by ayan-biswas0412** [comment]
> @ayan-biswas0412 Also did some work on this

Yeah here are the two POC videos for Google Classroom integration (which I did in my pre-GSoC period 😅),

1. https://youtu.be/UPnXXuXWHng
2. https://youtu.be/UPnXXuXWHng

But we need a clear cut flow for integrating Google Classroom for our CircuitVerse.

Challenges for Google Classroom are,
1. It is not open-sourced so we don't know how its internal thing works, and we neither make a plugin that can be installed in Google Classroom side to create assignments (interactive with CircuitVerse API) nor it supports LTI.
2. Possible pathways which I found at that time,
         I. Create the Google Classroom share button like shown in my video and also what @koolgax99 did.
         II. Conditionally enable the feature in the simulator part also such that students can submit their work to Google 
             Classroom, but again public projects can only be shared.
3. More research is needed for the following factors,
        I. Possible database field addition in projects/assignments to accommodate Google Classroom integration.
        II. Grade submission process for Classroom assignments.
        
Looking for ideas. Please join the channel  [#lms-integration](https://circuitverse-team.slack.com/archives/C01RBPJ570U) for more discussion regarding this.

cc @koolgax99 @tachyons @pulkit-30 @Prerna-0202

**Comment by koolgax99** [comment]
> > @ayan-biswas0412 Also did some work on this
> 
> Yeah here are the two POC videos for Google Classroom integration (which I did in my pre-GSoC period 😅),
> 
> 1. https://youtu.be/UPnXXuXWHng
> 2. https://youtu.be/UPnXXuXWHng
> 
> But we need a clear cut flow for integrating Google Classroom for our CircuitVerse.
> 
> Challenges for Google Classroom are,
> 
> 1. It is not open-sourced so we don't know how its internal thing works, and we neither make a plugin that can be installed in Google Classroom side to create assignments (interactive with CircuitVerse API) nor it supports LTI.
> 2. Possible pathways which I found at that time,
>    I. Create the Google Classroom share button like shown in my video and also what @koolgax99 did.
>    II. Conditionally enable the feature in the simulator part also such that students can submit their work to Google
>    Classroom, but again public projects can only be shared.
> 3. More research is needed for the following factors,
>    I. Possible database field addition in projects/assignments to accommodate Google Classroom integration.
>    II. Grade submission process for Classroom assignments.
> 
> Looking for ideas. Please join the channel [#lms-integration](https://circuitverse-team.slack.com/archives/C01RBPJ570U) for more discussion regarding this.
> 
> cc @koolgax99 @tachyons @pulkit-30 @Prerna-0202

For the grade submission process i guess there's no work around to get the gclassroom marks into the circuitverse circuit/database. The grade submission for the gclassroom assignment is independent. But do we want to achieve anything extra rather than just submitting the assignment?

**Comment by ayan-biswas0412** [comment]
> > > @ayan-biswas0412 Also did some work on this
> > 
> > 
> > Yeah here are the two POC videos for Google Classroom integration (which I did in my pre-GSoC period 😅),
> > 
> > 1. https://youtu.be/UPnXXuXWHng
> > 2. https://youtu.be/UPnXXuXWHng
> > 
> > But we need a clear cut flow for integrating Google Classroom for our CircuitVerse.
> > Challenges for Google Classroom are,
> > 
> > 1. It is not open-sourced so we don't know how its internal thing works, and we neither make a plugin that can be installed in Google Classroom side to create assignments (interactive with CircuitVerse API) nor it supports LTI.
> > 2. Possible pathways which I found at that time,
> >    I. Create the Google Classroom share button like shown in my video and also what @koolgax99 did.
> >    II. Conditionally enable the feature in the simulator part also such that students can submit their work to Google
> >    Classroom, but again public projects can only be shared.
> > 3. More research is needed for the following factors,
> >    I. Possible database field addition in projects/assignments to accommodate Google Classroom integration.
> >    II. Grade submission process for Classroom assignments.
> > 
> > Looking for ideas. Please join the channel [#lms-integration](https://circuitverse-team.slack.com/archives/C01RBPJ570U) for more discussion regarding this.
> > cc @koolgax99 @tachyons @pulkit-30 @Prerna-0202
> 
> For the grade submission process i guess there's no work around to get the gclassroom marks into the circuitverse circuit/database. The grade submission for the gclassroom assignment is independent. But do we want to achieve anything extra rather than just submitting the assignment?

regarding grade submission, I saw these last year but the flow is complicated https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions 
In the LTI case we have implemented two things 1. sharing assignment to LMS and the grade syncing with LMS but if we do Google Classroom then also I think we should implement these two

**Comment by pulkit-30** [comment]
After going through all the references and Google classroom API docs I have come up with these methods and a summary about Google Classroom Integration with CV, 

As of now, Circuitverse provides us the LMS facilities to :
- Create a group
- Add members in the Group
- Open Assignments
- Grade students on the basis of their performance and return them back.

A similar task we want to achieve with google classroom
- User can Start a new Course from the CV app
- Teacher can open assignments, questions, announcements, etc. from the CV app
- When students submit their assignments, the Teacher can Grade their performance and return their assignments back, that's of course from the CV app.

---

### Adding a classroom share Button 

https://user-images.githubusercontent.com/76155456/145054684-0d669443-3284-4a12-b154-e33755fd0fb7.mov

with the reference video,
we can integrate the google classroom by adding a share to the classroom button as also mentioned by @koolgax99 and @ayan-biswas0412 

reference: https://developers.google.com/classroom/guides/sharebutton

---

### By using Google Classroom Rest API

One more way, by which we can achieve these tasks is by integrating Google Classroom REST API with the CV app.

reference: https://developers.google.com/classroom/reference/rest

- To create a new Course: https://developers.google.com/classroom/reference/rest/v1/courses/create
- To get the list of the Courses: https://developers.google.com/classroom/reference/rest/v1/courses/list
- To create new announcement by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.announcements/create
- To create new Work Material: https://developers.google.com/classroom/reference/rest/v1/courses.courseWorkMaterials/create
- To add a User as a Student: https://developers.google.com/classroom/reference/rest/v1/courses.students/create
- To create a Teacher of a course: https://developers.google.com/classroom/reference/rest/v1/courses.teachers/create
- To open assignments by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork
- To get the Student Submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/get
- To return a student submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/return
- To Grade the Studnet assignment (by the Course teacher):https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions

#### Visit these links and try to execute the API (in try this API section) to know what the response we get and what are the parameters we need to specify to get the desired response.

Example :

https://user-images.githubusercontent.com/76155456/145056090-f916b0b7-0867-423c-8790-2cabd73fab2b.mov

This task is quite big, we can break down this task into small and start working on this.
I’m eager to receive your feedback @tachyons @ayan-biswas0412 @koolgax99 @Prerna-0202.

**Comment by ayan-biswas0412** [comment]
> After going through all the references and Google classroom API docs I have come up with these methods and a summary about Google Classroom Integration with CV,
> 
> As of now, Circuitverse provides us the LMS facilities to :
> 
> * Create a group
> * Add members in the Group
> * Open Assignments
> * Grade students on the basis of their performance and return them back.
> 
> A similar task we want to achieve with google classroom
> 
> * User can Start a new Course from the CV app
> * Teacher can open assignments, questions, announcements, etc. from the CV app
> * When students submit their assignments, the Teacher can Grade their performance and return their assignments back, that's of course from the CV app.
> 
> ### Adding a classroom share Button 
>  Screen.Recording.2021-12-07.at.12.16.25.AM.mov 
> with the reference video, we can integrate the google classroom by adding a share to the classroom button as also mentioned by @koolgax99 and @ayan-biswas0412
> 
> reference: https://developers.google.com/classroom/guides/sharebutton
> 
> ### By using Google Classroom Rest API
> One more way, by which we can achieve these tasks is by integrating Google Classroom REST API with the CV app.
> 
> reference: https://developers.google.com/classroom/reference/rest
> 
> * To create a new Course: https://developers.google.com/classroom/reference/rest/v1/courses/create
> * To get the list of the Courses: https://developers.google.com/classroom/reference/rest/v1/courses/list
> * To create new announcement by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.announcements/create
> * To create new Work Material: https://developers.google.com/classroom/reference/rest/v1/courses.courseWorkMaterials/create
> * To add a User as a Student: https://developers.google.com/classroom/reference/rest/v1/courses.students/create
> * To create a Teacher of a course: https://developers.google.com/classroom/reference/rest/v1/courses.teachers/create
> * To open assignments by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork
> * To get the Student Submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/get
> * To return a student submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/return
> * To Grade the Studnet assignment (by the Course teacher):https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions
> 
> #### Visit these links and try to execute the API (in try this API section) to know what the response we get and what are the parameters we need to specify to get the desired response.
> Example :
> 
>  Screen.Recording.2021-12-07.at.8.09.24.PM.mov 
> This task is quite big, we can break down this task into small and start working on this. I’m eager to receive your feedback @tachyons @ayan-biswas0412 @koolgax99 @Prerna-0202.

@pulkit-30 Share to Classroom approach I think is not apt according to our current flow of assignments because 
after sharing an assignment to GClassroom Circuitverse is not able to track the events.

Possible events we need to sync is 
- Assignment creation in Circuitverse end should create the assignment in Google Classroom
- Assignment grading from Circuitverse should update the Grade in Google Classroom
- To submit an assignment students can submit them via visiting CircuitVerse by clicking on the link in Google Classroom (as you shared in the video)

I think these things need to be brainstormed before the implementation of Google Classroom.

**Comment by pulkit-30** [comment]
> @pulkit-30 Share to Classroom approach I think is not apt according to our current flow of assignments because after sharing an assignment to GClassroom Circuitverse is not able to track the events.
> 
> Possible events we need to sync is
> 
> * Assignment creation in Circuitverse end should create the assignment in Google Classroom
> * Assignment grading from Circuitverse should update the Grade in Google Classroom
> * To submit an assignment students can submit them via visiting CircuitVerse by clicking on the link in Google Classroom (as you shared in the video)
> 
> I think these things need to be brainstormed before the implementation of Google Classroom.

When I first start searching for a better approach to implement this, I tried with the share to GClassroom Button, and also mentioned by you and @koolgax99 ...
But talking about assignment flow, Share to the classroom doesn't apt, that's why I go through the REST API references and check how it works and read more about it, and also mentioned some of the API requests above.
And yes we need to gather more information and brainstorm before the implementation of Google Classroom.

**Comment by koolgax99** [comment]
The important point here is to understand the use case and the usage/credibility of it. I feel, we cant do the same thing as google classroom does and ask the user to start using cv its completely not apt in terms of productivity, costing or other miscellaneous reasons. What we have to think if how to make it feasible for the CV student or a professor to manage their assignments via Google Classroom. In order to achieve this we need to brainstorm on the idea development of the GC integration such that A. It is feasible for both students and professors B. Not repeating the functionalities of GC instead finding a way to make it more apt. 

@tachyons @ayan-biswas0412 @pulkit-30 @Prerna-0202 please share your views.

**Comment by tachyons** [comment]
@ayan-biswas0412  Can you facilitate a brainstorming session for this? A time that works for @ayan-biswas0412 @pulkit-30 and @Prerna-0202 ?

Also don't worry about the credit of the work, git does support co-author feature so that every one who involved will get the credit for the work.

**Comment by ayan-biswas0412** [comment]
> @ayan-biswas0412 Can you facilitate a brainstorming session for this? A time that works for @ayan-biswas0412 @pulkit-30 and @Prerna-0202 ?
> 
> Also don't worry about the credit of the work, git does support co-author feature so that every one who involved will get the credit for the work.

sure, suggest some of the time slots @tachyons , sat/sun (except 19th Dec 2021) fine for me.

**Comment by tachyons** [comment]
> sure, suggest some of the time slots @tachyons , sat/sun (except 19th Dec 2021) fine for me.

I may not be available to join, I suggest you to the lead on this

**Comment by ayan-biswas0412** [comment]
> > @ayan-biswas0412 Can you facilitate a brainstorming session for this? A time that works for @ayan-biswas0412 @pulkit-30 and @Prerna-0202 ?
> > Also don't worry about the credit of the work, git does support co-author feature so that every one who involved will get the credit for the work.
> 
> sure, suggest some of the time slots @tachyons , sat/sun (except 19th Dec 2021) fine for me.

@ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then ? if yes can you share some free time slots (20-30 minutes will work I think).

**Comment by pulkit-30** [comment]
> @ayan-biswas0412 Can you facilitate a brainstorming session for this? A time that works for @ayan-biswas0412 @pulkit-30 and @Prerna-0202 ?
> 
> Also don't worry about the credit of the work, git does support co-author feature so that every one who involved will get the credit for the work.

Great Idea!

**Comment by pulkit-30** [comment]
> @ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then ? if yes can you share some free time slots (20-30 minutes will work I think).

Tomorrow or 18th Dec will be fine for me

**Comment by Prerna-0202** [comment]
> > @ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then ? if yes can you share some free time slots (20-30 minutes will work I think).
> 
> Tomorrow or 18th Dec will be fine for me

yes, Tomorrow or 18th Dec will be fine...

**Comment by ayan-biswas0412** [comment]
> > > @ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then? if yes can you share some free time slots (20-30 minutes will work I think)?
> > 
> > 
> > Tomorrow or 18th Dec will be fine for me
> 
> yes, Tomorrow or 18th Dec will be fine...

Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??

**Comment by pulkit-30** [comment]
> Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??

Yes, looks good!
whatsay @Prerna-0202

**Comment by Prerna-0202** [comment]
> > Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??
> 
> Yes, looks good! whatsay @Prerna-0202 5:30pm..

> > > > @ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then? if yes can you share some free time slots (20-30 minutes will work I think)?
> > > 
> > > 
> > > Tomorrow or 18th Dec will be fine for me
> > 
> > 
> > yes, Tomorrow or 18th Dec will be fine...
> 
> Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??

Yes, 5:30pm is fine...

**Comment by ayan-biswas0412** [comment]
> > > Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??
> > 
> > 
> > Yes, looks good! whatsay @Prerna-0202 5:30pm..
> 
> > > > > @ayan-biswas0412 @pulkit-30 @Prerna-0202 do you guys want to do a brainstorming session then? if yes can you share some free time slots (20-30 minutes will work I think)?
> > > > 
> > > > 
> > > > Tomorrow or 18th Dec will be fine for me
> > > 
> > > 
> > > yes, Tomorrow or 18th Dec will be fine...
> > 
> > 
> > Tomorrow 5:30 PM ?? @Prerna-0202 @pulkit-30 ??
> 
> Yes, 5:30pm is fine...

I have posted the Gmeet link in the `#community` channel in slack, please have a look,I look forward to meet you guys at 5:30 pm on 12 Dec 2021.

**Comment by ayan-biswas0412** [comment]
So what we discussed was to integrate the CircuitVerse with Classroom according to below 

Teacher Creates assignment in CircuitVerse -> Classroom checkbox ticked -> POST req to Classroom (handle google login in between)

And during Grade submission , (handle google login) -> Send Grades to Google Classroom 

@pulkit-30 please add your detailed plan here on this.

**Comment by pulkit-30** [comment]
> So what we discussed was to integrate the CircuitVerse with Classroom according to below
> 
> Teacher Creates assignment in CircuitVerse -> Classroom checkbox ticked -> POST req to Classroom (handle google login in between)
> 
> And during Grade submission , (handle google login) -> Send Grades to Google Classroom
> 
> @pulkit-30 please add your detailed plan here on this.

Sure, I will share all the details here by tomorrow.

**Comment by pulkit-30** [comment]
## LMS

A Learning Management System (LMS) is an online integrated software used for creating, delivering, tracking, and reporting educational courses and outcomes.

### 🔴 Google classroom is the third-party LMS integration.

* * *

## What we have in CV

Currently, in circuitverse we can
-\> Make a group
-\> Add members in that group
-\> Open an assignment in it
-\> Member complete that assignment and submit that.
-\> Mentor can check the assignment and grade the student on the basis of their performance.


* * *

## What we want ?
#### ( to integrate the google classroom with CV )

#### Approach -

- Using the share to classroom button
- REST API resources


## Rest API Resources :

We have given some API endpoints and we can send requests to these endpoints and responses from the API  will have the associated data.

link: https://developers.google.com/classroom/reference/rest

- To create a new Course: https://developers.google.com/classroom/reference/rest/v1/courses/create

* * *

- To get the list of the Courses: https://developers.google.com/classroom/reference/rest/v1/courses/list

* * *

- To create new announcement by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.announcements/create

`request body`
```
{
  "creatorUserId": "",
  "alternateLink": "",
  "id": "",
  "materials": [
    {
      "link": {
        "url": ""
      }
    }
  ],
  "text": "Sample Announcement",
  "state": "DRAFT",
  "courseId": "",
  "assigneeMode": "ALL_STUDENTS",
  "creationTime": "2021-12-12T15:01:23Z",
  "updateTime": "2021-12-12T15:01:23Z",
  "scheduledTime": "2021-12-12T20:33:23Z"
}
```

* * *

- To create new Work Material: https://developers.google.com/classroom/reference/rest/v1/courses.courseWorkMaterials/create

`request body`
```
{
  "creatorUserId": "",
  "alternateLink": "",
  "id": "",
  "materials": [
    {
      "link": {
        "url": ""
      }
    }
  ],
  "state": "PUBLISHED",
  "courseId": "",
  "assigneeMode": "ALL_STUDENTS",
  "creationTime": "2021-12-12T15:01:23Z",
  "updateTime": "2021-12-12T15:01:23Z",
  "title": "Sample Course Work",
  "description": "This is a sample couse work",
  "topicId": ""
}
```

* * *

- To add a User as a Student: https://developers.google.com/classroom/reference/rest/v1/courses.students/create

* * *

- To create a Teacher of a course: https://developers.google.com/classroom/reference/rest/v1/courses.teachers/create

* * *

- To open assignments by teachers for the students: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork

* * *

- To get the Student Submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/get

* * *

- To return a student submission: https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/return

* * *

- To Grade the Student assignment (by the Course teacher):https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions

* * *

## Posting An assignment (flow)

>Teacher Create a new assignment in CircuitVerse -> Classroom checkbox ticked -> POST req to Classroom (handle google login in between)

When Mentor creates an assignment in Circuitverse we have a checkbox above the `create assignment` button and when the user clicks on the `create assignment`  button and the checkbox is ticked then it will be `PUBLISHED` on Google Classroom as well as on Circuitverse.

`Post` request : https://classroom.googleapis.com/v1/courses/{courseId}/courseWork

<img width="1680" alt="Screenshot 2021-12-12 at 9 26 40 PM" src="https://user-images.githubusercontent.com/76155456/145809609-a3519456-f0c9-4fcf-ac44-8cca53d8411c.png">


* * *

## Handling the Grading Process

>And during Grade submission, (handle google log in) -> Send Grades to Google Classroom

We can use the `Patch` request to update the Grades of a Student
`Patch` request :https://classroom.googleapis.com/v1/courses/{courseId}/courseWork/{id}
`Patch` request :https://classroom.googleapis.com/v1/courses/{courseId}/courseWork/{courseWorkId}/studentSubmissions/{id}

link: 
https://developers.google.com/classroom/reference/rest/v1/courses.courseWork.studentSubmissions/patch


To make a request to the Google Classroom API we need 
- Google Auth2.0
- and the Google Classroom API key

Let's take an example:
A user signIn with his Google account, then he can directly make a request to the Google Classroom API,  but when the user is not SignIn with Google Auth 2.0 and he wants to create an assignment on the Google Classroom platform then he has to sign in with Google Account to make a request.

We need to figure out a way, for a user who is not signed in with Google. How he will make a request to Google Classroom.


Cc: @tachyons @ayan-biswas0412 @Prerna-0202

**Reply by Malavi1** [reply]
Great idea @pulkit-30

**Comment by pulkit-30** [comment]
hey @ayan-biswas0412,
I have added the details, 
Any add-on, or anything I forgot to add or any suggestion, kindly mention that.

**Comment by ayan-biswas0412** [comment]
> hey @ayan-biswas0412, I have added the details, Any add-on, or anything I forgot to add or any suggestion, kindly mention that.

@pulkit-30 can you please remove the Share to classroom thing from your above comments by editing and also removing the first video it will look more neat and clean then

**Comment by tachyons** [comment]
I enabled [discussions](https://github.com/CircuitVerse/CircuitVerse/discussions) on this repo, Can we move discussions ot there ?

**Comment by ayan-biswas0412** [comment]
> I enabled [discussions](https://github.com/CircuitVerse/CircuitVerse/discussions) on this repo, Can we move discussions ot there ?

sure

**Comment by pulkit-30** [comment]
Hey @tachyons @ayan-biswas0412 @Prerna-0202,
Please share any updates or progress so far ... so that we can continue working on this!

**Reply by ayan-biswas0412** [reply]
Yeah, actually as we discussed this last time, you can go ahead and make one proof of concept of the google classroom integration.
Don't think too much about the UI for now, you can just make a simple working flow of the integration between Google Classroom and CircuitVerse.

@Prerna-0202 and @pulkit-30 both of you can jointly do this

**Comment by Prerna-0202** [comment]
Sorry, I was busy with my college stuff 😔 ........
